### PR TITLE
Sort games table in UI

### PR DESCRIPTION
**Pull request by MariusTulbure**
Fixes https://github.com/robert-damoc/game-stats/issues/21

image

### CODE REVIEW

**Review by robert-damoc — COMMENTED** [review]
I forgot to add some details to the ticket. We also want to have the direction. The default will be `id: :asc`. Will update the ticket to include the new details

**Comment by MariusTulbure** [comment]
image

image

image

File: app/controllers/games_controller.rb [hunk 1]
```diff
@@ -3,7 +3,12 @@ class GamesController < ApplicationController
 
   # GET /games or /games.json
   def index
-    @games = Game.all
+    allowed_sort_columns = %w[id created_at state]
+
+    sort_column = params[:sort].presence_in(allowed_sort_columns) || :id
+    sort_order = params[:order] == 'desc' ? 'desc' : 'asc'
```
Comment (robert-damoc) [inline app/controllers/games_controller.rb:9]: I don't think this is the place for these variables. I will add more details in the review comment

File: app/views/games/_gametable.html.erb [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
+
+
+<table class="table table table-striped table-hover tabletable-sm">
+  <thead class="table-dark table-bordered ">
     <tr style="text-align: center">
-      <th scope="col">Number</th>
-      <th scope="col">State</th>
-      <th scope="col">Created</th>
+      <th>
+        <%= link_to "Game Number", "?sort=id&order=#{params[:sort] == 'id' && params[:order] == 'asc' ? 'desc' : 'asc'}", class:'plain-link' %>
```
Comment (robert-damoc) [inline app/views/games/_gametable.html.erb:7]: `"?sort=id&order=#{params[:sort] == 'id' && params[:order] == 'asc' ? 'desc' : 'asc'}"` This part looks a bit hard to understand. I'd rather use hashes to pass the params, as
```erb
        <%= link_to "State", {
          sort: 'state',
          sort_dir: @sort == 'state' && @sort_dir == 'asc' ? 'desc' : 'asc'
        }, class:'plain-link' %>
```

**Review by robert-damoc — COMMENTED** [review]
I think this looks good. But there is a better and cleaner approach here. The sorting functionality will be shared across multiple places, so it'd be good to already take steps on creating a good base for that.

You already did the first step, in creating a component for the arrow. We can expend that by moving somewhere shared - right now it is a partial for games, but we can create something like `views/shared/_sorting_arrow.html.erb`, and have that `views/shared/` be the base of our common things for UI,

The sorting login you have in the controller, should be part of the module actually. But better than adding it in all modules that need it (duplicate code), or having it in the `application_record.rb` (which means that models that don't need it will have it), we should create a concern. Something like `app/models/concerns/sortable.rb`, that would look like this:
```ruby
module Sortable
  extend ActiveSupport::Concern

  class ColumnNotSortableBy < StandardError; end
  class UndefinedDirection < StandardError; end

  class_methods do
    ALLOWED_DIRECTIONS = %w[asc desc]

    def sort_table(column_name = nil, direction = nil)
      column_name ||= default_sort_column
      direction ||= default_direction

      raise ColumnNotSortableBy, "Can't sort by #{column_name}" unless allowed_sort_columns.include?(column_name)
      raise UndefinedDirection, "Undefined direction: #{direction}" unless ALLOWED_DIRECTIONS.include?(direction)

      order(column_name => direction)
    end

    def default_sort_column
      :id
    end

    def default_direction
      :asc
    end
  end
end
```

With something like this, we can easily include it in all the modules that need it, and also overwrite the default sorting and direction, plus defining the columns we want to allow sorting on.

For the controller, it's good to define 2 iVariables, like `@sort` and `@sort_dir`, to be able to use them in views, where their values will be the param[:sort] || default value (same for dir)

File: app/views/games/_game_table.html.erb [hunk 3]
```diff
@@ -1,16 +1,17 @@
-<table class="table table table-striped table-hover table-sm">
-  <thead class="table-dark">
+
```
Comment (robert-damoc) [inline app/views/games/_game_table.html.erb:1]: Empty extra line here

File: app/views/shared/_table_header.html.erb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<th>
+  <%= link_to header_name, {
+    sort: sort_key,
+    sort_dir: @sort == sort_key && @sort_dir == 'asc' ? 'desc' : 'asc'
+  }, class: 'plain-link' %>
+  <%if @sort == sort_key %>
```
Comment (robert-damoc) [inline app/views/shared/_table_header.html.erb:6]:
```suggestion
  <% if @sort == sort_key %>
```

**Review by robert-damoc — APPROVED** [review]

**Pull request merged**